DOC: Replace Sphinx search with Algolia

bjnath opened the pull request:
'Quick search' removed from sidebars; all pages now have a searchbar at the top. Sphinx search remains accessible in a dedicated page.

This is the search used in pytorch.org (among [many others](https://docsearch.algolia.com/)).

I'll post the CI artifact to demonstrate the look and feel. However,  WIP -- always says "no results found."  The PR gets a live version up for Algolia to take a look at.

Closes gh-17037.

<!-- Please be sure you are following the instructions in the dev guidelines
http://www.numpy.org/devdocs/dev/development_workflow.html
-->

<!-- We'd appreciate it if your commit message is properly formatted
http://www.numpy.org/devdocs/dev/development_workflow.html#writing-the-commit-message
-->

bjnath committed:
DOC: Add Algolia searchbar

Replaces Sphinx search. 'Quick search' removed from sidebars.

mattip commented:
xref issue gh-17037

I think we should only change the search engine in this PR and not touch the formatting. One thing at a time.

bjnath commented:
There's no formatting change. This is based on master. New logo went in with #16988.

bjnath commented:
Artifact: https://15086-908607-gh.circle-artifacts.com/0/doc/build/html/index.html

mattip commented on doc/source/_static/numpy.css:
```diff
@@ -0,0 +1,32 @@
+the search results get clipped by the sidebar margin */
+#searchbox {
+  visibility: hidden;
+}
```
This seems like a formatting change, no?

bjnath replied:
This is the Sphinx "Quick search" searchbar that's being hidden, as noted above:

> 'Quick search' removed from sidebars

This is done so pages won't have two unrelated searchbars in different locations.

eric-wieser commented on doc/source/_templates/layout.html:
```diff
@@ -1,35 +1,22 @@
 {% extends "!layout.html" %}
 
 {%- block extrahead %}
-<style>
-  .main {
-    -moz-box-shadow: none;
-    -webkit-box-shadow: none;
-    box-shadow: none;
-  }
-  div.top-scipy-org-logo-header {
-    background-color: #fafafa;
-    border-bottom: 2px solid #013243; /* Warm Black */
-    margin-top: 0;
-    box-shadow: none;
-  }
-  div.top-scipy-org-logo-header img {
-    height: 100px;
-    padding-left: 27px;
-  }
-  div.spc-navbar .nav-pills > li > a {
-    background-color: #4d77cf; /* Han Blue */
-  }
-</style>
```
Can you move this code around in a standalone PR?

bjnath replied:
This code is already in #16988. I've moved it out of the header. This is not two PRs in one.

eric-wieser replied:
I'm suggesting making a PR that does nothing but move this code. It makes your other PRs conflict less, and makes all the PRs easier to review.

bjnath replied:
This makes sense and I'll bear it in mind for future PRs.

eric-wieser commented on doc/source/_templates/layout.html:
```diff
@@ -1,35 +1,22 @@
 {%- block header %}
 <div class="container">
   <div class="top-scipy-org-logo-header">
-    <a href="{{ pathto('index') }}">
```
This looks like an unrelated change too

bjnath replied:
That is true.

eric-wieser commented on doc/source/_templates/layout.html:
```diff
@@ -52,3 +39,17 @@
+<script type="text/javascript"> docsearch({
+apiKey: 'c8367ae0f87be4fd19cb01bae97ef4f2',
+indexName: 'numpy',
+inputSelector: 'form input[type=text]',
```
This seems dangerously vague - can we use a more specific filter?
```suggestion
inputSelector: '#algolia-searchbar form input[type=text]',
```

bjnath replied:
Agreed.

bashtage commented:
This implementation is different from how pytorch works.  In pytorch, the equivalent search box on this page:

https://15086-908607-gh.circle-artifacts.com/0/doc/build/html/reference/index.html

uses standard sphinx search.

bjnath committed:
DOC: Add suggested changes to PR #17056

Addresses reviewer comments in
https://github.com/numpy/numpy/pull/17056/files#r468547323 and
https://github.com/numpy/numpy/pull/17056/files#r468547781

bjnath commented:
I've changed the PR title and comments to clarify that Sphinx search is still there and still usable.

bashtage commented:
I don't see how this is opt-in.  I would think that opt-in would require checking an environmental variable like `NUMPY_USE_ALGOLIA` and this code if 1 (or true, or equivalent).

rgommers commented:
> uses standard sphinx search.
> 
> Mirroring pytorch would be placing this search on the numpy.org website, not in the reference documentation.

This is a good point. It does make sense to put this on `numpy.org` only. If we do that, I'd like to unify the main docs and NEPs builds again though. Separate deployment of http://www.numpy.org/neps/ never made much sense to me, and having the NEPs unsearchable from the docs is quite annoying.

bjnath commented:
@rgommers 
> It does make sense to put this on numpy.org only.

I think users will be baffled by having search engines change underneath them. They have no reason to expect different results on different pages.

bjnath commented:
> opt-in would require checking an environmental variable like NUMPY_USE_ALGOLIA 

Without yet considering the exact mechanism, does everyone see configurability an acceptable solution?

mattip commented:
My concerns about adding Javascript that sends information to a commercial entity still remain

rgommers commented:
> Without yet considering the exact mechanism, does everyone see configurability as an acceptable solution?

No, that doesn't make much sense to me. 

> My concerns about adding Javascript that sends information to a commercial entity still remain

If there's a better alternative I'd love to hear it, but my prio 1 is to actually _have_ a search box that covers all our content.

rgommers commented:
We also have Google Analytics by the way, that sends data to a commercial entity ....

bjnath commented:
> that sends information to a commercial entity

Rather than battling this a PR at a time, with possibly inconsistent outcomes, we should have a policy in place. 

Submitted issue #17066

mattip commented:
> We also have Google Analytics by the way, that sends data to a commercial entity ....
> And Cloudflare, that keeps stats on all our visitors ....

We should definitely reconsider Google Analytics. More lightweight and transparent alternatives exist. I don't think anyone actually considered the implications when adding it, and a quick search will show that many people have issues with it.

I don't know much about data collection via CloudFare, but that is a different story: they actually provide a useful caching service.

bjnath committed:
DOC: Fix Algolia javascript in #17056

Removes erroneous 'algoliaOptions' args.

bjnath commented:
The search engine is working at https://15101-908607-gh.circle-artifacts.com/0/doc/build/html/index.html

Even if you think these guys are devils and would never put Algolia on our site, you it to yourself to have a look.

Content from NEPs and numpy.org are included.

bashtage commented:
How do you get a paginate search if you need to look beyond the top 5?  I couldn't get it to work on mobile.

Also, it is duplicating the results. For example, try entering the start of concatenate.

bjnath commented:
@bashtage
> Also, it is duplicating the results. 

One is from devdocs, one is from stable.  Making the choice of dev/stable user-selectable so you see only one or the other.

bjnath commented:
> How do you get a paginate search if you need to look beyond the top 5?

The search on their site has scrollbars; I've asked why ours doesn't.

charris commented:
Ping all.

mattip commented:
Let's close this in the hopes that I will finish gh-17265 soonish.
